src/benchmarks/micro/libraries/System.IO.FileSystem/Perf.FileStream.cs

```diff
@@ -2,152 +2,299 @@
+                while (bytesRead < fileSize)
                 {
-                    bytesRead += await reader.ReadAsync(buffer, 0, buffer.Length);
+                    bytesRead += fileStream.Read(userBuffer, 0, userBuffer.Length);
                 }
             }
 
             return bytesRead;
         }
 
         [Benchmark]
-        [BenchmarkCategory(Categories.NoWASM)]
-        public async Task CopyToAsync()
+        [Arguments(OneKiloByte, HalfKiloByte, FileOptions.None)]
+        [Arguments(OneKiloByte, FourKiloBytes, FileOptions.None)]
+        [Arguments(OneMegaByte, HalfKiloByte, FileOptions.None)]
+        [Arguments(OneMegaByte, FourKiloBytes, FileOptions.None)]
+        [Arguments(HundredMegaBytes, HalfKiloByte, FileOptions.None)]
+        [Arguments(HundredMegaBytes, FourKiloBytes, FileOptions.None)]
+        public void Write(long fileSize, int userBufferSize, FileOptions options)
         {
-            using (var reader = new FileStream(_filePath, FileMode.Open, FileAccess.Read, FileShare.Read, DefaultBufferSize, FileOptions.Asynchronous))
+            byte[] userBuffer = _userBuffers[userBufferSize];
+            using (FileStream fileStream = new FileStream(_destinationFilePaths[fileSize], FileMode.Create, FileAccess.Write, FileShare.Read, FourKiloBytes, options))
             {
-                await reader.CopyToAsync(Stream.Null);
+                for (int i = 0; i < fileSize / userBufferSize; i++)
+                {
+                    fileStream.Write(userBuffer, 0, userBuffer.Length);
+                }
             }
         }
 
         [Benchmark]
-        public void WriteByte()
+        [Arguments(OneKiloByte, HalfKiloByte, FileOptions.Asynchronous)]
+        [Arguments(OneKiloByte, FourKiloBytes, FileOptions.Asynchronous)]
+        [Arguments(OneMegaByte, HalfKiloByte, FileOptions.Asynchronous)]
+        [Arguments(OneMegaByte, FourKiloBytes, FileOptions.Asynchronous)]
+        [Arguments(HundredMegaBytes, HalfKiloByte, FileOptions.Asynchronous)]
+        [Arguments(HundredMegaBytes, FourKiloBytes, FileOptions.Asynchronous)]
+        [BenchmarkCategory(Categories.NoWASM)]
+        public async Task<long> ReadAsync(long fileSize, int userBufferSize, FileOptions options)
         {
-            using (FileStream writer = new FileStream(_filePath, FileMode.Create, FileAccess.Write, FileShare.Read, DefaultBufferSize, FileOptions.None))
+            byte[] userBuffer = _userBuffers[userBufferSize];
+            long bytesRead = 0;
+            using (FileStream fileStream = new FileStream(_sourceFilePaths[fileSize], FileMode.Open, FileAccess.Read, FileShare.Read, FourKiloBytes, options))
             {
-                for (int i = 0; i < TotalSize; i++)
+                while (bytesRead < fileSize)
                 {
-                    writer.WriteByte(default);
+                    bytesRead += await fileStream.ReadAsync(userBuffer, 0, userBuffer.Length);
```

Comment:
If we're only going to choose one overload, I'd prefer it be the one that takes a Memory.  Same for Write.

Reply:
Perhaps we should have both of them? The array overload allows for apples to apples comparison with Full Framework, which sometimes is nice to have. And it's also most probably the most frequently used overload?

Reply:
> The array overload allows for apples to apples comparison with Full Framework

Does that matter?

> And it's also most probably the most frequently used overload?

It might be today.  It shouldn't be in the future.  For example, we've changed almost all use of Stream.ReadAsync(byte[], ...) throughout the core libraries where we're consuming some stream to be Stream.ReadAsync(Memory, ...), and [CA1835](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1835) will recommend to developers that they switch overloads and provides an auto-fixer to do so.

If someone cares about performance, they should be using these newer overloads.  Today ReadAsync(Memory, ...) enables zero-allocation for synchronously completing operations, and I'm hopeful that with the .NET 6 FileStream work that will be extended to both ReadAsync and WriteAsync for asynchronously completion as well (ammortized zero-allocation).  They enable pre-pinned memory to be used and avoid the GCHandle costs that are otherwise required.  And they enable reading/writing with non-array-based sources, potentially saving on copying costs to get things into arrays.